pdns/slavecommunicator.cc

```diff
@@ -48,20 +48,34 @@
 
 #include "ixfr.hh"
 
-void CommunicatorClass::addSuckRequest(const DNSName &domain, const ComboAddress& master, bool force)
+void CommunicatorClass::addSuckRequest(const DNSName &domain, const ComboAddress& master, bool force, uint8_t priority)
 {
   std::lock_guard<std::mutex> l(d_lock);
   SuckRequest sr;
   sr.domain = domain;
   sr.master = master;
   sr.force = force;
+  sr.priorityAndOrder.first = priority;
+  sr.priorityAndOrder.second = d_sorthelper++;
   pair<UniQueue::iterator, bool>  res;
 
-  res=d_suckdomains.push_back(sr);
+  res=d_suckdomains.insert(sr);
   if(res.second) {
     d_suck_sem.post();
+  } else {
+    // Domain is already in there, we should check the priority and whether its forced
+    domains_by_name_t& nameindex=boost::multi_index::get<IDTag>(d_suckdomains);
```

Comment:
Just a nit but if I'm not mistaken you could use `project()` (`d_suckdomains.project<IDTag>(res.first);`) to avoid doing a second lookup.

Reply:
I realized I didn't even need the project, I could just call modify on the original iter directly.